drivers/rtc/rtc_ifx_cat1.c

```diff
@@ -0,0 +1,338 @@
+}
+
+static int ifx_cat1_rtc_init(const struct device *dev)
+{
+	cy_rslt_t rslt = CY_RSLT_SUCCESS;
+	static const cy_stc_rtc_config_t default_time = {
+		.dayOfWeek = CY_RTC_SATURDAY,
+		.date = 1,
+		.month = 1,
+		.year = 0,
+	};
+
+	Cy_SysClk_ClkBakSetSource(CY_SYSCLK_BAK_IN_CLKLF);
+
+	if (_ifx_cat1_rtc_get_state() == _IFX_CAT1_RTC_STATE_UNINITIALIZED) {
+		if (Cy_RTC_IsExternalResetOccurred()) {
+			/* Reset to default time */
+			Cy_RTC_SetDateAndTime(&default_time);
+			_ifx_cat1_rtc_set_century(_IFX_CAT1_RTC_INIT_CENTURY);
+		}
+
+		if (Cy_SysPm_RegisterCallback(&_ifx_cat1_rtc_pm_cb)) {
+			_ifx_cat1_rtc_set_state(_IFX_CAT1_RTC_STATE_ENABLED);
+		} else {
+			rslt = -EINVAL;
+		}
+	} else if (_ifx_cat1_rtc_get_state() == _IFX_CAT1_RTC_STATE_ENABLED ||
+		   _ifx_cat1_rtc_get_state() == _IFX_CAT1_RTC_STATE_TIME_SET) {
+
+		if (Cy_RTC_GetInterruptStatus() & CY_RTC_INTR_CENTURY) {
+			_ifx_cat1_rtc_century_interrupt();
+		}
+	}
+
+	Cy_RTC_ClearInterrupt(CY_RTC_INTR_CENTURY);
+	Cy_RTC_SetInterruptMask(CY_RTC_INTR_CENTURY);
+
+	_ifx_cat1_rtc_dst = NULL;
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), _ifx_cat1_rtc_isr_handler,
+		    DEVICE_DT_INST_GET(0), 0);
+	irq_enable(DT_INST_IRQN(0));
+
+	return rslt;
+}
+
+static int ifx_cat1_rtc_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	struct ifx_cat1_rtc_data *data = dev->data;
+
+	uint32_t sec = timeptr->tm_sec;
+	uint32_t min = timeptr->tm_min;
+	uint32_t hour = timeptr->tm_hour;
+	uint32_t day = timeptr->tm_mday;
+	uint32_t mon = timeptr->tm_mon + 1;
+	uint32_t year = timeptr->tm_year + _IFX_CAT1_RTC_TM_YEAR_BASE;
+	uint32_t year2digit = year % 100;
+
+	cy_rslt_t rslt;
+	int err = 0;
+	uint32_t retry = 0;
+
+	if (!CY_RTC_IS_SEC_VALID(sec) || !CY_RTC_IS_MIN_VALID(min) || !CY_RTC_IS_HOUR_VALID(hour) ||
+	    !CY_RTC_IS_MONTH_VALID(mon) || !CY_RTC_IS_YEAR_SHORT_VALID(year2digit) ||
+	    !IFX_CAT1_RTC_VALID_CENTURY(year)) {
+
+		return -EINVAL;
+	}
+	do {
+		if (retry != 0) {
+			_IFX_CAT1_RTC_WAIT_ONE_MS();
+		}
+
+		err = k_mutex_lock(&data->lock, K_NO_WAIT);
+		if (err) {
+			return err;
+		}
+
+		rslt = Cy_RTC_SetDateAndTimeDirect(sec, min, hour, day, mon, year2digit);
+		if (rslt == CY_RSLT_SUCCESS) {
+			_ifx_cat1_rtc_set_century((uint16_t)(year) - (uint16_t)(year2digit));
+		}
+
+		k_mutex_unlock(&data->lock);
+		++retry;
+	} while (rslt == CY_RTC_INVALID_STATE && retry < _IFX_CAT1_RTC_MAX_RETRY);
+
+	retry = 0;
+	while (CY_RTC_BUSY == Cy_RTC_GetSyncStatus() && retry < _IFX_CAT1_RTC_MAX_RETRY) {
+		_IFX_CAT1_RTC_WAIT_ONE_MS();
+		++retry;
+	}
+
+	if (rslt == CY_RSLT_SUCCESS) {
+		_ifx_cat1_rtc_set_state(_IFX_CAT1_RTC_STATE_TIME_SET);
+	}
+	return rslt;
+}
+
+static int ifx_cat1_rtc_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	struct ifx_cat1_rtc_data *data = dev->data;
+	int err = 0;
+
+	cy_stc_rtc_config_t dateTime = {.hrFormat = CY_RTC_24_HOURS};
+
+	err = k_mutex_lock(&data->lock, K_NO_WAIT);
```

Comment:
please use a spinlock or binary semaphore for locking instead of a mutex. The mutex is quite "expensive", tracking ownership of the lock etc. which is not needed for a simple get/put :)

```
k_sem_take(&data->lock, K_NO_WAIT);
k_sem_give(&data->lock);
```